### PR TITLE
Support for Data Contracts

### PR DESCRIPTION
**Pull request by rayokota**
- Add support for data contracts
- Add support for CSFLE for AWS, Azure, GCP, and Hashicorp Vault
- Add support for CEL, CEL_FIELD, and JSONata rules

### CODE REVIEW

File: src/confluent_kafka/schema_registry/schema_registry_client.py [hunk 1]
```diff
@@ -670,7 +867,7 @@ def get_versions(self, subject_name):
 
         return self._rest_client.get('subjects/{}/versions'.format(_urlencode(subject_name)))
 
-    def delete_version(self, subject_name, version):
+    def delete_version(self, subject_name: str, version: int) -> int:
```
Comment (Claimundefine) [inline src/confluent_kafka/schema_registry/schema_registry_client.py:870]: There's no cache invalidation for delete version?

File: src/confluent_kafka/schema_registry/schema_registry_client.py [hunk 2]
```diff
@@ -372,26 +545,41 @@ class SchemaRegistryClient(object):
         `Confluent Schema Registry documentation <http://confluent.io/docs/current/schema-registry/docs/intro.html>`_
     """  # noqa: E501
 
-    def __init__(self, conf):
+    def __init__(self, conf: dict):
+        self._conf = conf
         self._rest_client = _RestClient(conf)
         self._cache = _SchemaCache()
```
Comment (Claimundefine) [inline src/confluent_kafka/schema_registry/schema_registry_client.py:551]: Is there any case where we would want TTL for schemaCache and RegisteredSchemaCache?

File: src/confluent_kafka/schema_registry/schema_registry_client.py [hunk 3]
```diff
@@ -111,34 +112,119 @@ def __init__(self, conf):
...
+        self.auth = userinfo if userinfo != ('', '') else None
+
+        # The following adds support for proxy config
+        # If specified: it uses the specified proxy details when making requests
+        self.proxy = None
+        proxy = conf_copy.pop('proxy', None)
+        if proxy is not None:
+            self.proxy = proxy
+
+        self.timeout = None
```
[the remark this answers is not in the document]
Comment (rayokota) [reply src/confluent_kafka/schema_registry/schema_registry_client.py:124]: Good catch.  I'll change it.

File: src/confluent_kafka/schema_registry/schema_registry_client.py [hunk 4]
```diff
@@ -372,26 +545,41 @@ class SchemaRegistryClient(object):
...
     def __exit__(self, *args):
         if self._rest_client is not None:
-            self._rest_client._close()
+            self._rest_client.session.close()
+
+    def config(self):
+        return self._conf
 
-    def register_schema(self, subject_name, schema, normalize_schemas=False):
+    def register_schema(
```
[the remark this answers is not in the document]
Comment (rayokota) [reply src/confluent_kafka/schema_registry/schema_registry_client.py:572]: Good catch!  I'll add it

File: src/confluent_kafka/schema_registry/schema_registry_client.py [hunk 5]
```diff
@@ -586,31 +755,73 @@ def get_latest_version(self, subject_name):
...
+            fmt (str): Format of the schema
+
+        Returns:
+            RegisteredSchema: Registration information for this version.
+
+        Raises:
+            SchemaRegistryError: if the version can't be found or is invalid.
+        """  # noqa: E501
+
+        cache_key = (subject_name, frozenset(metadata.items()))
```
[the remark this answers is not in the document]
Comment (rayokota) [reply src/confluent_kafka/schema_registry/schema_registry_client.py:793]: Another good one, I'll add it

File: src/confluent_kafka/schema_registry/schema_registry_client.py (hunk 1 above)
Comment (rayokota) [reply src/confluent_kafka/schema_registry/schema_registry_client.py:870]: Not currently, I can take a look if it's easy to add

File: src/confluent_kafka/schema_registry/schema_registry_client.py (hunk 2 above)
Comment (rayokota) [reply src/confluent_kafka/schema_registry/schema_registry_client.py:551]: Not in normal scenarios

**Comment by rayokota** [comment]
Thanks @Claimundefine , I've incorporated your feedback

**Review by Claimundefine — APPROVED** [review]
LGTM

**Review by Claimundefine — APPROVED** [review]

**Pull request merged**